### ISSUE TITLE
Hash tables can be made 2× or 4× smaller

### ISSUE THREAD

**Issue by greatroar** [pierrec/lz4#92]
CompressBlock uses a hash table of 1<<16 int-typed match positions, which is 256kiB on a 32-bit platform, 512kiB on a 64-bit one. That's wasteful, since match offsets are limited to 65536 by the LZ4 format. If the table entry type is changed to uint16, it uses only 128kiB, regardless of the platform.

The trick is to store only the lower 16 bits of the match position, meaning that it becomes an offset past an implicit 64kiB block boundary. Since the compressor doesn't remember where that boundary was, it has to check for every match in the hash table whether an actual match occurs at si-offset. That's not too much of a problem, because it has to check that position anyway in case of a hash collision and it's likely still in the cache (more likely when the hash table is smaller).

I have a [working prototype](/greatroar/lz4/tree/smaller-hashtables) of this idea. It passes all tests and its speed is largely unchanged on amd64:

```
name              old time/op    new time/op    delta
Compress-8          3.13ms ± 3%    3.05ms ± 1%    -2.42%  (p=0.000 n=20+20)
CompressRandom-8    17.9µs ± 1%     8.4µs ± 1%   -52.72%  (p=0.000 n=19+20)
CompressPg1661-8     685µs ± 2%     700µs ± 2%    +2.26%  (p=0.000 n=20+20)
CompressDigits-8     652µs ± 3%     649µs ± 2%      ~     (p=0.659 n=20+20)
CompressTwain-8      677µs ± 4%     690µs ± 3%    +1.90%  (p=0.000 n=20+20)
CompressRand-8       641µs ± 3%     634µs ± 2%    -1.16%  (p=0.006 n=20+19)
```

Results on the Raspberry PI look worse, but because the allocation rate has high variance, I think the sync.Pools don't work as well on that tiny machine.

I'd like to clean up the code and submit a PR, but before I do so, I'd like to discuss the API with you. I think the current CompressBlock API doesn't work all too well anymore, given the type change. The prototype code casts a []int hash table to *[1<<16]uint16 with an unsafe hack. If I may make a proposal, I think a new API might look as follows:

```go
// A BlockCompressor compresses byte slices into the LZ4 block format.
type BlockCompressor struct {/* ... */}

func (c *BlockCompressor) Compress(src, dst []byte) (int, error) // Or just int? #91

type BlockCompressor struct {
    Level int
    // implementation
}

func (c *BlockCompressor) Compress(src, dst []byte) (int, error)
```

Having these types allows us to get rid of the sync.Pools. The old functions can be compatibility shims that call these new ones:

```go
func CompressBlock(src, dst []byte, _ []int) (int, error) {
    var c BlockCompressor // Maybe use a sync.Pool for this one.
    return c.Compress(src, dst)
}
```

WDYT?

**Comment by pierrec** [comment]
This idea makes sense.
What about the Uncompress functions? Do you propose another type or a new method?

**Comment by greatroar** [comment]
Uncompress would not change, since it's stateless. I can try to set a PR soonish.

**Comment by greatroar** [comment]
Fixed by 0004c74bab2e34a90c2045722116673cd8318497.
